### ISSUE TITLE
Option to SFTP Connection Script

### ISSUE THREAD

**Issue by daltenbern3** [microsoft/vscode-remote-release#5906]
Issue #4129 For reference:
[https://github.com/microsoft/vscode-remote-release/issues/4129](https://github.com/microsoft/vscode-remote-release/issues/4129)

Due to issues with CyberArk truncating the initialization script, it is impossible to use the Remote Development extension when connecting to a host through CyberArk.

I would suggest an option that allows the connection script to be sent via SFTP to the remote host then executed afterward. This would force the user to type their credentials in twice but would only be enabled if the user so chooses. 

This would allow users who are locked behind a CyberArk Proxy Server to be able to use this extension, while not changing the experience for any standard users.

**Comment by singsang2** [comment]
I have the same problem!

**Comment by jscott88** [comment]
This is definitely and issue. Hoping that something can be done about it soon.

**Comment by hlouidor** [comment]
Finally! Hopefully someone will look into this!

**Comment by bdub7** [comment]
Agreed.

**Comment by go4colin** [comment]
Let's make this happen.

**Comment by nbily1** [comment]
Very happy this made it into the backlog! Seems it's been raised a few times, but never got enough support to go anywhere. 

Just gathering context from the previous issues raised with both the Remote Dev team and CyberArk.

# Issue

The default way VS Code passes the install script to the remote host is to pipe the script (as a string?) to the SSH command.

CyberArk PSM uses a jump server to handle the username/password authentication and then connects to the target server from there.

Because the install script is piped, the jump server attempts to run it before the connection has been established to the target (see fig. 1). There are a couple reasons this won't work:
1. Running the install script on the jump server is no good because the target server won't be set up correctly
2. The jump box has some file size limit (4KB?) that causes the script to get truncated and throws and "unexpected EOF" error

### Figure 1
![image](https://user-images.githubusercontent.com/94550179/143266670-82e7a059-3581-4714-9342-4f2010c62d29.png)

# Proposed solutions

## Add a delay before running the install script

I'm thinking this ultimately won't work because the jump box will still try to execute everything piped to the SSH command before attempting to connect to the target box.

## Use SCP to pass the install script instead of piping it

Something like this:

### Figure 2
![image](https://user-images.githubusercontent.com/94550179/143267697-4adcdb2c-6df4-4d28-9f0c-6abc0e02eb5a.png)

Potential drawback is that the user may need to authenticate twice. Personally, this is better than not being able to connect at all. And if this were an option in settings.json, it would not affect users that do not need to use this method. However, using an SFTP client, I am freely able to pass files back and forth without having to authenticate every time (but maybe this is somehow different).

**Comment by olegasdo** [comment]
Hello
Any update on the progress of the issue?

**Comment by wgm32** [comment]
I, and several other developers and data scientists, are having this same issue right now when trying to develop remotely on a CyberArk managed server.  Has there been any further action on this request to make the install script over sftp a configurable option?  Maybe have an option for the user to run the script one time after the first connection?  In this day and time of heightened cybersecurity practices and zero trust many more organizations are likely to be using CyberArk as part of their cybersecurity posture.  This issue is likely to be encountered more frequently.

**Comment by amey2688** [comment]
i agree with @wgm32 . Many companies which require highly secure infrastructure are using cyberark. due to this issue , people in companies using cyber ark will not be able to do remote development.this issue should be fixed

**Comment by ckapolonaris** [comment]
Any news anyone? I'm dealing with the same issues as the rest of the data scientists above

**Comment by jorgenspange** [comment]
Any updates here?

**Comment by waardd** [comment]
I also would like to know if there is any progress here

**Comment by dijarvrella** [comment]
A solution here would be greatly appreciated.

**Comment by joshspicer** [comment]
Does this cyberark tool permit copying files with `scp`?  

That solution wasn't available back when this issue was created, but now it can. Copying via scp is automatic as a fallback, but you can force it with `remote.SSH.localServerDownload = always`.  This [comment](remote.SSH.localServerDownload) seems to suggest scp is supported, but I don't have access to cyberark to confirm this.

**Comment by ifoughal** [comment]
Has this being fixed?
